### PR TITLE
Journalist has to be logged in to create an article

### PR DESCRIPTION
**Pull request by salindholm**
[Journalist has to be logged in to create an article](https://www.pivotaltracker.com/story/show/176339828)
## User Story
```
As an API,
In order to keep the content secure,
I would like to be able to have the journalists to log in before creating articles
```

## Changes proposed in this pull request:
* Adding User model with devise_token_auth
* Creating a test and factory bot for User model
* Adding role to User model using enum
* Creating authentication and authorization for create action in article controller
* Creating a test to check if only journalist can create an article
* Adding a route to create articles

## What I have learned working on this feature:
* The importance of running all tests to see if the new implementation breaks something
* How to test for enums

## Packages/Gems added
* devise_token_auth

### CODE REVIEW

File: app/controllers/api/articles_controller.rb [hunk 1]
```diff
@@ -25,4 +28,10 @@ def create
   def article_params
     params.require(:article).permit(:title, :lead, :body, :category_id)
   end
+
+  def is_user_journalist?
+    unless current_user.journalist?
+      render json: { message: 'You are not authorized to create an article.' }, status: 401
+    end
```
Comment (tochman) [inline app/controllers/api/articles_controller.rb:35]: You can use this now, but I would prefer if you used Pundit for learning purposes. NOT a blocker though

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,3 @@
 class ApplicationController < ActionController::API
+        include DeviseTokenAuth::Concerns::SetUserByToken
```
Comment (tochman) [inline app/controllers/application_controller.rb]: Please format your code

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+RSpec.describe User, type: :model do
+  it 'is expected to have valid Factory' do
+    expect(create(:user)).to be_valid
+    expect(create(:registered_user, email: 'registered@example.com')).to be_valid
+    expect(create(:journalist, email: 'journalist@example.com')).to be_valid
+  end
```
Comment (tochman) [inline spec/models/user_spec.rb:12]: These expect statements should all be in separate `it` blocks/tests.

File: spec/requests/api/journalist_can_create_an_article_spec.rb [hunk 4]
```diff
@@ -87,4 +95,47 @@
...
+               category_id: category.id,
+               author_id: 'Test author'
+             }
+           }
+    end
+    it 'is expected to return 401 status' do
+      expect(response).to have_http_status 401
+    end
+    it 'is expected to return an error message' do
+      expect(response_json['errors']).to eq ['You need to sign in or sign up before continuing.']
```
Comment (tochman) [inline spec/requests/api/journalist_can_create_an_article_spec.rb]: Not sure if this is the message you want to send, but more importantly, please stick to the same standards you use above in this expect statement. `have_key` and `have_value` are good...

File: app/controllers/api/articles_controller.rb (hunk 1 above)
Comment (salindholm) [reply app/controllers/api/articles_controller.rb:35]: Okey noted, we will implement it next time!

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (salindholm) [reply app/controllers/application_controller.rb]: Fixed

File: spec/models/user_spec.rb (hunk 3 above)
Comment (salindholm) [reply spec/models/user_spec.rb:12]: Fixed!

File: spec/requests/api/journalist_can_create_an_article_spec.rb (hunk 4 above)
Comment (salindholm) [reply spec/requests/api/journalist_can_create_an_article_spec.rb]: Good input, we have now corrected tests for better consistency!

**Pull request merged**